File: frontend/apps/ios/Pantopus/Features/Shared/Feed/FeedComponents.swift
//
//  FeedComponents.swift
//  Pantopus
//
//  Reusable building blocks for feed surfaces (Pulse today, Gigs next).
//  Each piece is a thin SwiftUI view that consumes a small content
//  model — feature code assembles them to match the design.
//

import SwiftUI

// MARK: - Chip row

/// One entry in `FeedChipRow`. The `id` is the filter key sent to the
/// view-model; the `label` is the display string.
public struct FeedChipItem: Identifiable, Sendable, Hashable {
    public let id: String
    public let label: String

    public init(id: String, label: String) {
        self.id = id
        self.label = label
    }
}

/// Horizontal scrolling chip row used by Pulse + Gigs to filter the
/// feed in place. The active chip uses the primary fill; idle chips
/// have a thin border.
public struct FeedChipRow: View {
    private let chips: [FeedChipItem]
    private let activeId: String
    private let onSelect: @MainActor (String) -> Void

    public init(
        chips: [FeedChipItem],
        activeId: String,
        onSelect: @escaping @MainActor (String) -> Void
    ) {
        self.chips = chips
        self.activeId = activeId
        self.onSelect = onSelect
    }

    public var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: Spacing.s2) {
                ForEach(chips) { chip in
                    let active = chip.id == activeId
                    Button { onSelect(chip.id) } label: {
                        Text(chip.label)
                            .font(.system(size: 12.5, weight: .semibold))
                            .foregroundStyle(active ? Theme.Color.appTextInverse : Theme.Color.appTextStrong)
                            .padding(.horizontal, 14)
                            .frame(height: 28)
                            .background(active ? Theme.Color.primary600 : Theme.Color.appSurface)
                            .overlay(
                                RoundedRectangle(cornerRadius: Radii.pill, style: .continuous)
                                    .stroke(active ? .clear : Theme.Color.appBorder, lineWidth: 1)
                            )
                            .clipShape(RoundedRectangle(cornerRadius: Radii.pill, style: .continuous))
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel(chip.label)
                    .accessibilityAddTraits(active ? [.isButton, .isSelected] : .isButton)
                    .accessibilityIdentifier("feedChip_\(chip.id)")
                }
            }
            .padding(.horizontal, Spacing.s4)
            .padding(.vertical, Spacing.s3)
        }
        .background(Theme.Color.appBg)
        .overlay(alignment: .bottom) {
            Rectangle()
                .fill(Theme.Color.appBorder)
                .frame(height: 1)
        }
        .accessibilityIdentifier("feedChipRow")
    }
}

// MARK: - Compose FAB

/// Pencil FAB used by feed surfaces to start a compose flow.
public struct FeedComposeFAB: View {
    private let action: @MainActor () -> Void
    private let accessibilityLabel: String

    public init(
        accessibilityLabel: String = "Compose post",
        action: @escaping @MainActor () -> Void
    ) {
        self.accessibilityLabel = accessibilityLabel
        self.action = action
    }

    public var body: some View {
        Button(action: action) {
            Icon(.pencil, size: 20, strokeWidth: 2.2, color: Theme.Color.appTextInverse)
                .frame(width: 52, height: 52)
                .background(Theme.Color.primary600)
                .clipShape(Circle())
                .shadow(color: Theme.Color.primary600.opacity(0.36), radius: 12, x: 0, y: 8)
        }
        .buttonStyle(.plain)
        .accessibilityLabel(accessibilityLabel)
        .accessibilityIdentifier("feedComposeFAB")
    }
}

// MARK: - Skeleton card

/// Single shimmer card used by the loading frame. Optionally renders an
/// extra title-line shimmer for intents whose cards have a title
/// (Event today; Gigs uses titles too).
public struct FeedSkeletonCard: View {
    private let withTitle: Bool

    public init(withTitle: Bool = false) {
        self.withTitle = withTitle
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: Spacing.s2) {
            HStack(spacing: 9) {
                Shimmer(width: 32, height: 32, cornerRadius: Radii.xl)
                VStack(alignment: .leading, spacing: 5) {
                    Shimmer(width: 110, height: 10, cornerRadius: Radii.xs)
                    Shimmer(width: 70, height: 8, cornerRadius: Radii.xs)
                }
                Spacer(minLength: Spacing.s2)
                Shimmer(width: 42, height: 16, cornerRadius: Radii.pill)
            }
            if withTitle {
                Shimmer(height: 11, cornerRadius: Radii.xs)
                    .frame(maxWidth: 200)
            }
            Shimmer(height: 9, cornerRadius: Radii.xs)
            Shimmer(height: 9, cornerRadius: Radii.xs)
                .frame(maxWidth: .infinity)
            HStack(spacing: 14) {
                Shimmer(width: 56, height: 10, cornerRadius: Radii.xs)
                Shimmer(width: 56, height: 10, cornerRadius: Radii.xs)
                Spacer()
                Shimmer(width: 42, height: 10, cornerRadius: Radii.xs)
            }
        }
        .padding(Spacing.s3)
        .background(Theme.Color.appSurface)
        .clipShape(RoundedRectangle(cornerRadius: Radii.lg, style: .continuous))
        .overlay(
            RoundedRectangle(cornerRadius: Radii.lg, style: .continuous)
                .stroke(Theme.Color.appBorder, lineWidth: 1)
        )
        .accessibilityHidden(true)
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/shared/feed/FeedComponents.kt
@file:Suppress("MagicNumber", "PackageNaming")

package app.pantopus.android.ui.screens.shared.feed

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.pantopus.android.ui.components.Shimmer
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.PantopusIconImage
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing

/** One entry in [FeedChipRow]. */
@Immutable
data class FeedChipItem(
    val id: String,
    val label: String,
)

/** Horizontal scrolling chip row used by Pulse + Gigs to filter in place. */
@Composable
fun FeedChipRow(
    chips: List<FeedChipItem>,
    activeId: String,
    onSelect: (String) -> Unit,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier =
            modifier
                .fillMaxWidth()
                .background(PantopusColors.appBg)
                .testTag("feedChipRow"),
    ) {
        Row(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .horizontalScroll(rememberScrollState())
                    .padding(horizontal = Spacing.s4, vertical = Spacing.s3),
            horizontalArrangement = Arrangement.spacedBy(Spacing.s2),
        ) {
            chips.forEach { chip ->
                val active = chip.id == activeId
                Box(
                    modifier =
                        Modifier
                            .height(28.dp)
                            .clip(RoundedCornerShape(Radii.pill))
                            .background(if (active) PantopusColors.primary600 else PantopusColors.appSurface)
                            .then(
                                if (active) {
                                    Modifier
                                } else {
                                    Modifier.border(1.dp, PantopusColors.appBorder, RoundedCornerShape(Radii.pill))
                                },
                            )
                            .clickable { onSelect(chip.id) }
                            .padding(horizontal = 14.dp)
                            .testTag("feedChip_${chip.id}"),
                    contentAlignment = Alignment.Center,
                ) {
                    Text(
                        text = chip.label,
                        fontSize = 12.5.sp,
                        fontWeight = FontWeight.SemiBold,
                        color = if (active) PantopusColors.appTextInverse else PantopusColors.appTextStrong,
                    )
                }
            }
        }
        Box(
            modifier =
                Modifier
                    .align(Alignment.BottomCenter)
                    .fillMaxWidth()
                    .height(1.dp)
                    .background(PantopusColors.appBorder),
        )
    }
}

/** Pencil FAB used by feed surfaces to start a compose flow. */
@Composable
fun FeedComposeFAB(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    contentDescription: String = "Compose post",
) {
    Box(
        modifier =
            modifier
                .size(52.dp)
                .shadow(elevation = 12.dp, shape = CircleShape)
                .clip(CircleShape)
                .background(PantopusColors.primary600)
                .clickable(onClick = onClick)
                .testTag("feedComposeFAB")
                .semantics { this.contentDescription = contentDescription },
        contentAlignment = Alignment.Center,
    ) {
        PantopusIconImage(
            icon = PantopusIcon.Pencil,
            contentDescription = null,
            size = Radii.xl2,
            tint = PantopusColors.appTextInverse,
        )
    }
}

/**
 * Single shimmer card used by the loading frame. Optionally renders an
 * extra title-line shimmer for intents whose cards have a title.
 */
@Composable
fun FeedSkeletonCard(
    withTitle: Boolean = false,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier =
            modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(Radii.lg))
                .border(1.dp, PantopusColors.appBorder, RoundedCornerShape(Radii.lg))
                .background(PantopusColors.appSurface)
                .padding(Spacing.s3),
        verticalArrangement = Arrangement.spacedBy(Spacing.s2),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(9.dp)) {
            Shimmer(width = 32.dp, height = 32.dp, cornerRadius = Radii.xl)
            Column(modifier = Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(5.dp)) {
                Shimmer(width = 110.dp, height = 10.dp, cornerRadius = Radii.xs)
                Shimmer(width = 70.dp, height = 8.dp, cornerRadius = Radii.xs)
            }
            Shimmer(width = 42.dp, height = 16.dp, cornerRadius = Radii.pill)
        }
        if (withTitle) {
            Shimmer(width = 200.dp, height = 11.dp, cornerRadius = Radii.xs)
        }
        Shimmer(width = 320.dp, height = 9.dp, cornerRadius = Radii.xs)
        Shimmer(width = 280.dp, height = 9.dp, cornerRadius = Radii.xs)
        Row(horizontalArrangement = Arrangement.spacedBy(14.dp), verticalAlignment = Alignment.CenterVertically) {
            Shimmer(width = 56.dp, height = 10.dp, cornerRadius = Radii.xs)
            Shimmer(width = 56.dp, height = 10.dp, cornerRadius = Radii.xs)
            Spacer(modifier = Modifier.weight(1f))
            Shimmer(width = 42.dp, height = 10.dp, cornerRadius = Radii.xs)
        }
    }
}

/** Helper that builds the standard padding around the feed content area. */
fun feedContentPadding() = PaddingValues(Spacing.s3)
